app/src/main/java/org/wikipedia/dataclient/WikiSite.kt

```diff
@@ -0,0 +1,211 @@
+            langVariant
+        } else {
+            authorityToLanguageCode(authority)
+        }
+
+        // This prevents showing mixed Chinese variants article when the URL is /zh/ or /wiki/ in zh.wikipedia.org
+        if (languageCode == AppLanguageLookUpTable.CHINESE_LANGUAGE_CODE) {
+            languageCode = firstSelectedChineseVariant
+        }
+
+        // Use default subdomain in authority to prevent error when requesting endpoints. e.g. zh-tw.wikipedia.org
+        if (authority.contains("wikipedia.org") && subdomain().isNotEmpty()) {
+            authority = subdomain() + ".wikipedia.org"
+        }
+        this.uri = Uri.Builder().scheme(tempUri.scheme).encodedAuthority(authority).build()
+    }
+
+    constructor(url: String) : this(
+        when {
+            url.startsWith("http") -> Uri.parse(url)
+            url.startsWith("//") -> Uri.parse("$DEFAULT_SCHEME:$url")
+            else -> Uri.parse("$DEFAULT_SCHEME://$url")
+        }
+    )
+
+    constructor(authority: String, languageCode: String) : this(authority) {
+        this.languageCode = languageCode
+    }
+
+    fun scheme(): String {
+        return if (uri.scheme.isNullOrEmpty()) DEFAULT_SCHEME else uri.scheme!!
+    }
+
+    /**
+     * @return The complete wiki authority including language subdomain but not including scheme,
+     * authentication, port, nor trailing slash.
+     *
+     * @see [URL syntax](https://en.wikipedia.org/wiki/Uniform_Resource_Locator.Syntax)
+     */
+    fun authority(): String {
+        return uri.authority.orEmpty()
+    }
+
+    fun subdomain(): String {
+        return languageCodeToSubdomain(languageCode)
+    }
+
+    /**
+     * @return A path without an authority for the segment including a leading "/".
+     */
+    fun path(segment: String): String {
+        return "/w/$segment"
+    }
+
+    /**
+     * @return The canonical URL. e.g., https://en.wikipedia.org.
+     */
+    fun url(): String {
+        return uri.toString()
+    }
+
+    /**
+     * @return The canonical URL for segment. e.g., https://en.wikipedia.org/w/foo.
+     */
+    fun url(segment: String): String {
+        return url() + path(segment)
+    }
+
+    // TODO: this method doesn't have much to do with WikiSite. Move to PageTitle?
+    /**
+     * Create a PageTitle object from an internal link string.
+     *
+     * @param internalLink Internal link target text (eg. /wiki/Target).
+     * Should be URL decoded before passing in
+     * @return A [PageTitle] object representing the internalLink passed in.
+     */
+    fun titleForInternalLink(internalLink: String?): PageTitle {
+        // Strip the /wiki/ from the href
+        return PageTitle(removeInternalLinkPrefix(internalLink!!), this)
+    }
+
+    // TODO: this method doesn't have much to do with WikiSite. Move to PageTitle?
+    /**
+     * Create a PageTitle object from a Uri, taking into account any fragment (section title) in the link.
```

Comment:
This comment could probably be removed?

Reply:
What's the problem with it?

Reply:
We have just been deleting unnecessary comments in favor of self-commenting code during conversion thats all.